rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java

```diff
@@ -121,20 +124,31 @@ public BlockChainImpl loadBlockchain() {
         Block bestBlock = blockStore.getBestBlock();
         if (bestBlock == null) {
             logger.info("DB is empty - adding Genesis");
-            for (RskAddress addr : genesis.getPremine().keySet()) {
-                repository.createAccount(addr);
-                InitialAddressState initialAddressState = genesis.getPremine().get(addr);
-                repository.addBalance(addr, initialAddressState.getAccountState().getBalance());
-                AccountState accountState = repository.getAccountState(addr);
-                accountState.setNonce(initialAddressState.getAccountState().getNonce());
-
-                if (initialAddressState.getContractDetails()!=null) {
-                    repository.updateContractDetails(addr, initialAddressState.getContractDetails());
-                    accountState.setStateRoot(initialAddressState.getAccountState().getStateRoot());
-                    accountState.setCodeHash(initialAddressState.getAccountState().getCodeHash());
+
+            // first we need to create the accounts, which creates also the associated ContractDetails
+            for (RskAddress accounts : genesis.getAccounts().keySet()) {
+                repository.createAccount(accounts);
+            }
+
+            // second we create contracts whom only have code modifying the preexisting ContractDetails instance
+            for (Map.Entry<RskAddress, byte[]> codeEntry : genesis.getCodes().entrySet()) {
+                RskAddress contractAddress = codeEntry.getKey();
+                ContractDetails contractDetails = repository.getContractDetails(contractAddress);
+                contractDetails.setCode(codeEntry.getValue());
+                Map<DataWord, byte[]> contractStorage = genesis.getStorages().get(contractAddress);
+                for (Map.Entry<DataWord, byte[]> storageEntry : contractStorage.entrySet()) {
+                    byte[] storageValue = storageEntry.getValue();
+                    if (storageValue == null) {
```

Comment:
I wouldn't do kind of check, if the storage values are wrong is better to know when they are inserted. I would remove the if condition